include/dfu/dfu_target_smp.h

```diff
@@ -0,0 +1,142 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief DFU target reset callback for acticate recovery mode.
+ *
+ * @return 0 on success, negative errno otherwise.
+ */
+typedef int (*dfu_target_reset_cb_t)(void);
+
+/**
+ * @brief Register recovery mode reset callback
+ *
+ * Function call enable that DFU target will wake up Target device to Recovery mode
+ * before start send SMP commands.
+ *
+ * @param[in] cb User definided target reset function before enable recovery mode
+ *
+ * @return 0 on success, negative errno otherwise.
+ */
+int dfu_target_smp_recovery_mode_enable(dfu_target_reset_cb_t cb);
+
+/**
+ * @brief Initialize dfu target SMP client
+ *
+ * @retval 0 If successful, negative errno otherwise.
+ */
+int dfu_target_smp_client_init(void);
+
+/**
+ * @brief Read Image list
+ *
+ * @param res_buf Image List buffer
+ *
+ * @return 0 on success, negative errno otherwise.
+ */
+int dfu_target_smp_image_list_get(struct mcumgr_image_state *res_buf);
+
+/**
+ * @brief See if data in buf indicates MCUBoot style upgrade.
+ *
+ * @retval true if data matches, false otherwise.
+ */
+bool dfu_target_smp_identify(const void *const buf);
+
+/**
+ * @brief Initialize dfu target, perform steps necessary to receive firmware.
+ *
+ * @param[in] file_size Size of the current file being downloaded.
+ * @param[in] img_num Image pair index.
+ * @param[in] cb Callback for signaling events(unused).
+ *
+ * @retval 0 If successful, negative errno otherwise.
+ */
+int dfu_target_smp_init(size_t file_size, int img_num, dfu_target_callback_t cb);
+
+/**
+ * @brief Get offset of firmware
+ *
+ * @param[out] offset Returns the offset of the firmware upgrade.
+ *
+ * @return 0 if success, otherwise negative value if unable to get the offset
+ */
+int dfu_target_smp_offset_get(size_t *offset);
+
+/**
+ * @brief Write firmware data.
+ *
+ * @param[in] buf Pointer to data that should be written.
+ * @param[in] len Length of data to write.
+ *
+ * @return 0 on success, negative errno otherwise.
+ */
+int dfu_target_smp_write(const void *const buf, size_t len);
+
+/**
+ * @brief Deinitialize resources and finalize firmware upgrade if successful.
+
+ * @param[in] successful Indicate whether the firmware was successfully received.
+ *
+ * @return 0 on success, negative errno otherwise.
+ */
+int dfu_target_smp_done(bool successful);
+
+/**
+ * @brief Schedule update of one or more images.
+ *
+ * This call requests images update. The update will be performed after
+ * the device reset.
+ *
+ * @param[in] img_num Given image pair index or -1 for all
+ *		      of image pair indexes.
+ *
+ * @return 0 for a successful request or a negative error
+ *	   code identicating reason of failure.
+ **/
+int dfu_target_smp_schedule_update(int img_num);
+
+/**
+ * @brief Release resources and erase the download area.
+ *
+ * Cancels any ongoing updates.
+ *
+ * @return 0 on success, negative errno otherwise.
+ */
+int dfu_target_smp_reset(void);
```

Comment:
Not sure "reset" should be used for "erase" when there is also a "reset_command" function, should probably have erase in the name

Reply:
Ah let check this How other DFU target's behaviour.

Reply:
I think you have a correct point. Let me change this API and own API for "Reset command send"